Improving server exception logging by including exception in log message

valbers opened the pull request:
There were two problems before:

1. The error message itself was not being logged, but only a mention that there was an error;
2. It was not possible to access the possibly original exception which led to the request processing error. However, it's extremely useful to log the error with the whole exception information including the stack trace of where the exception was thrown.

### Before
<img width="674" alt="before-srv-ex-logging" src="https://github.com/fsprojects/FSharp.Data.GraphQL/assets/15154973/8c0d8a28-481e-40df-9b3f-25ada055d25f">


### After
![after-srv-ex-logging](https://github.com/fsprojects/FSharp.Data.GraphQL/assets/15154973/84f8bbc1-0eef-4a33-9706-2ba21eb0e9b9)


  However, these changes also introduce a breaking change as `GQLProblemDetails` now has an additional member holding the possible exception which generated the error.

   `GQLProblemDetails` is public and could be in use by code using FSharp.Data.GraphQL.

valbers committed:
.AspNetCore: BREAKING CHANGE: logging server exceptions with exception info too

  There were two problems before:
    1. The error message itself was not being logged, but only a mention
       that there was an error;
    2. It was not possible to access the possibly original exception
       which led to the request processing error. However, it's
       extremely useful to log the error with the whole exception
       information including the stack trace of where the exception was thrown.
  However, these changes also introduce a breaking change as IGQLError
and GQLProblemDetails now each have an additional mandatory member
holding the possible exception which generated the error.
  These types (especiall GQLProblemDetails) are public and could be used
by code using FSharp.Data.GraphQL.

valbers committed:
refactor: trying to improve readability by removing confusing active pattern

valbers committed:
adapting code to new exception parameter

valbers committed:
PART II of "adapting code to new exception parameter"

xperiandri commented on src/FSharp.Data.GraphQL.Shared/Errors.fs:
```diff
@@ -11,6 +11,7 @@ type internal FieldPath = obj list
 
 type IGQLError =
     abstract member Message : string with get
+    abstract member Exception : (Exception option) with get
```
Maybe we just introduce one more interface?
```
type IGQLErrorException =
    abstract member Exception : Exception voption with get
```

valbers replied:
But how would it be used? How would I access the exception later like I do when `IGQLError` has just another member? I think there can be only one interface in the end.

xperiandri replied:
Match if interface is implemented

valbers replied:
Done.

xperiandri commented on src/FSharp.Data.GraphQL.Shared/Errors.fs:
```diff
@@ -127,50 +134,57 @@ type GQLProblemDetails = {
         | _ -> ReadOnlyDictionary<string, obj> mutableExtensions
 
-    static member internal CreateWithKind (message, kind : ErrorKind, ?path) = {
+    static member internal CreateWithKind (message : string, ex : Exception option,  kind : ErrorKind, ?path) = {
```
Maybe we make them `voption` and `[<Optional>]`?
```suggestion
    static member internal CreateWithKind (message : string, kind : ErrorKind, [<Optional>] ex, [<Optional>] path) = {
```

valbers replied:
Fine by me.

valbers replied:
Although: I just rewrote my changes according to your suggestions regarding the extra interface and, with it, this change at `CreateWithKind` isn't necessary anymore.

xperiandri commented on src/FSharp.Data.GraphQL.Shared/SchemaDefinitions.fs:
If we introduce new interface this change will not be needed

valbers replied:
Ok... But I still don't get how introducing a new interface will then actually solve the problem. Where will this interface be used? 🤔

xperiandri replied:
Then we can introduce more overloads

xperiandri commented on tests/FSharp.Data.GraphQL.Tests/AstValidationTests.fs:
```diff
@@ -11,8 +11,8 @@ open FSharp.Data.GraphQL.Types
 
-    let CreateValidation message = GQLProblemDetails.CreateWithKind (message, Validation)
-    let CreateValidationFor path message = GQLProblemDetails.CreateWithKind (message, Validation, path)
+    let CreateValidation message = GQLProblemDetails.CreateWithKind (message, None, Validation)
+    let CreateValidationFor path message = GQLProblemDetails.CreateWithKind (message, None, Validation, path)
```
And with optionals we can place ex at the end and not specify here

valbers replied:
I thought about that possibility too (i.e.: just append the exception parameter to the list of parameters). But the reason why I put the exception argument in the middle is because the message and the exception are tightly related, almost as if they were two members of the same type. So it makes more sense to have them one next to the other than separated by other parameters.

valbers replied:
Not necessary anymore with the now extra `IGQLExceptionError` interface.

valbers committed:
Revert "PART II of "adapting code to new exception parameter""

This reverts commit 46d5b760d988fc2e45dac274495883f866c18856.

valbers committed:
Revert "adapting code to new exception parameter"

This reverts commit d24abb4b9331d120b1848daec637dad8b14f693e.

valbers committed:
Revert "refactor: trying to improve readability by removing confusing active pattern"

This reverts commit b03bf400d8bf19eedb8526db6ba86a4f3dbd9369.

valbers committed:
Revert ".AspNetCore: BREAKING CHANGE: logging server exceptions with exception info too"

This reverts commit e829181c9a0b7948529796f0142007dac53b2851.

valbers committed:
.AspNetCore: (2nd attempt): logging server exceptions with exception info too

  Note: this commit is a 2nd attempt after getting feedback from code
review.

  There were two problems before:
    1. The error message itself was not being logged, but only a mention
       that there was an error;
    2. It was not possible to access the possibly original exception
       which led to the request processing error. However, it's
       extremely useful to log the error with the whole exception
       information including the stack trace of where the exception was thrown.
  However, these changes also introduce a breaking change as IGQLError
and GQLProblemDetails now each have an additional mandatory member
holding the possible exception which generated the error.
  These types (especiall GQLProblemDetails) are public and could be used
by code using FSharp.Data.GraphQL.

valbers commented:
@xperiandri I rewrote my changes according to your suggestion regarding the extra interface. Please kindly review the current changes.

xperiandri commented on src/FSharp.Data.GraphQL.Server.AspNetCore/Giraffe/HttpHandlers.fs:
```diff
@@ -139,11 +139,23 @@ module HttpHandlers =
+                        ex,
+                        "Error while processing request that generated response with documentId '{documentId}'",
+                        documentId
+                    )
```
Shouldn't we just call `logger.LogError` as many times as how many exceptions we have?

valbers replied:
Done.

xperiandri reviewed (commented):
Just one suggestion

valbers committed:
.AspNetCore: for each possible exception logging it

  These changes are according to a request during code review.

valbers committed:
Removed duplicate fantomas config. value (was causing problems)

valbers committed:
.AspNetCore: formatting HttpHandlers with Fantomas

valbers commented:
@xperiandri there's a commit for applying fantomas formatting to `HttpHandlers.fs` but I find it suspicious. Do you have an idea why the formatting is so different now?

xperiandri commented on src/FSharp.Data.GraphQL.Shared/Errors.fs:
```diff
@@ -216,7 +257,15 @@ type GQLProblemDetails = {
-        GQLProblemDetails.Create (message, path, extensions)
+        match error with
+        | :? IGQLExceptionError as exceptionError ->
+            match extensions with
+            | Include x ->
+                GQLProblemDetails.CreateOfException(exceptionError.Message, exceptionError.Exception, path, extensions = x)
+            | Skip ->
+                GQLProblemDetails.CreateOfException(exceptionError.Message, exceptionError.Exception, path)
```
We can make it simpler. Anyway `option` is created under the hood
```suggestion
            let x = extensions |> Skippable.toOption
            GQLProblemDetails.CreateOfException(exceptionError.Message, exceptionError.Exception, path, ?extensions = x)
```

valbers replied:
This doesn't work. Believe me, I tried this 🙂 . It doesn't expect an `option`, although the parameter is `option` under the hood. It's apparently a characteristic of the `?extensions` syntax.

xperiandri replied:
I've made it work in my code but I implemented a better change

xperiandri replied:
See PR to your branch

xperiandri commented on src/FSharp.Data.GraphQL.Shared/Errors.fs:
```diff
@@ -200,7 +233,15 @@ type GQLProblemDetails = {
-        GQLProblemDetails.Create (message, path, extensions)
+        match error with
+        | :? IGQLExceptionError as exceptionError ->
+            match extensions with
+            | Include x ->
+                GQLProblemDetails.CreateOfException(exceptionError.Message, exceptionError.Exception, path, extensions = x)
+            | Skip ->
+                GQLProblemDetails.CreateOfException(exceptionError.Message, exceptionError.Exception, path)
```
The same as below

valbers replied:
It's not possible. See my previous comment.

valbers replied:
Done in your PR. Closing converstaion.

xperiandri commented on src/FSharp.Data.GraphQL.Shared/Errors.fs:
```diff
@@ -187,7 +212,15 @@ type GQLProblemDetails = {
-        GQLProblemDetails.Create (message, extensions)
+        match error with
+        | :? IGQLExceptionError as exceptionError ->
+            match extensions with
+            | Include x ->
+                GQLProblemDetails.CreateOfException(exceptionError.Message, exceptionError.Exception, extensions = x)
+            | Skip ->
+                GQLProblemDetails.CreateOfException(exceptionError.Message, exceptionError.Exception)
```
The same as below

valbers replied:
It's not possible. See my previous comment.

valbers replied:
Done in your PR. Closing comment

xperiandri reviewed (commented):
@valbers it is something bad on your side. I've just formatted it on my side and it is similar to what it was before your formatting.
Try restore the latest Fantomas with tool restore in this repo

xperiandri committed:
Reduced allocations in `GQLProblemDetails` creation

xperiandri committed:
Formatted `HttpHandlers` correctly

valbers committed:
Revert ".AspNetCore: formatting HttpHandlers with Fantomas"

This reverts commit 273796ad88025fcc9597d907589a3d46c6d786cc.

valbers commented:
@xperiandri waiting for your approving review.

xperiandri commented on src/FSharp.Data.GraphQL.Shared/Errors.fs:
```diff
@@ -129,48 +140,71 @@ type GQLProblemDetails = {
     }
 
-    static member Create (message, path, ?extensions : IReadOnlyDictionary<string, obj>) = {
+    static member CreateOfException (message : string, ex : Exception, [<Optional>] path : FieldPath Skippable, [<Optional>] extensions : IReadOnlyDictionary<string, obj> Skippable) = {
```
Shouldn't we call it just `Create`? As in my mind if you emphasize that it is `CreateOfException` it is something like you read the Exception fields into the new object.
While adding just an input parameter have no additional cognitive load

xperiandri replied:
@njlr @nikhedonia @cmeeren @pkese any thoughts?

valbers replied:
Ok. It's just that I tend to avoid overloads, because I like to have a glance at the code and know exactly which function/method is being called 😄 .
But yeah, I can rename it to `Create`. This way it blends in with the other overloads.

cmeeren replied:
I don't use this library yet, but sure, I definitely prefer overloads in these cases, for several reasons I don't care articulating at the moment.

valbers replied:
Done.

xperiandri reviewed (commented):
What do you think?

valbers committed:
"CreateOfException" -> "Create"

xperiandri commented on src/FSharp.Data.GraphQL.Server.AspNetCore/Giraffe/HttpHandlers.fs:
```diff
@@ -115,48 +106,59 @@ module HttpHandlers =
-                    documentId,
-                    metadata
-                )
+                if
+                    errs
+                    |> Seq.choose (fun x -> x.Exception |> Option.ofValueOption)
+                    |> (not << Seq.isEmpty)
+                then
+                    errs
+                    |> List.choose (fun x -> x.Exception |> Option.ofValueOption)
+                    |> List.iter (fun ex ->
+                        logger.LogError (ex, "Error while processing request that generated response with documentId '{documentId}'", documentId))
+                else
+                    logger.LogWarning (
+                        ("Produced request error GraphQL response with:\n"
+                         + "- documentId: '{documentId}'\n"
+                         + "- error(s):\n    {requestError}\n"
+                         + "- metadata:\n    {metadata}\n"),
+                        documentId,
+                        errs,
+                        metadata
+                    )
```
Maybe we just iter over once without allocating new lists?
```suggestion
                let notHasException =
                    errs
                    |> Seq.map (fun err ->
                        err.Exception |> ValueOption.iter (fun ex ->
                            logger.LogError (ex, "Error while processing request that generated response with documentId '{documentId}'", documented)
                        err.Exception.IsNone
                    |> Seq.fold (fun acc notExn -> acc && notExn) true
                if notHasException then
                    logger.LogWarning (
                        ("Produced request error GraphQL response with:\n"
                         + "- documentId: '{documentId}'\n"
                         + "- error(s):\n    {requestError}\n"
                         + "- metadata:\n    {metadata}\n"),
                        documentId,
                        errs,
                        metadata
                    )
```

xperiandri replied:
Probably fold can be replaced with `Seq.all id`

valbers replied:
I call this "the Tetris effect in functional programming" 😄

valbers replied:
Done.

xperiandri reviewed (commented):
One last thing 🙂

valbers committed:
.AspNetCore: more memory-efficient exception logging

  Co-authored-by: Andrii Chebukin <XperiAndri@Outlook.com>

xperiandri commented:
Cool! Lots of thanks for the contribution!
